File: app/src/main/java/app/kamy/qalbuApp/domain/tools/QiblaCalculator.kt
package app.kamy.qalbuApp.domain.tools

import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.tan

object QiblaCalculator {
    private const val KAABA_LAT = 21.4225
    private const val KAABA_LNG = 39.8262

    /** Bearing from user location to Kaaba in degrees (0–360, clockwise from north). */
    fun bearingToKaaba(latitude: Double, longitude: Double): Float {
        val lat1 = Math.toRadians(latitude)
        val lat2 = Math.toRadians(KAABA_LAT)
        val deltaLng = Math.toRadians(KAABA_LNG - longitude)
        val y = sin(deltaLng) * cos(lat2)
        val x = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(deltaLng)
        val bearing = Math.toDegrees(atan2(y, x))
        return ((bearing + 360) % 360).toFloat()
    }

    fun distanceToKaabaKm(latitude: Double, longitude: Double): Double {
        val earthRadius = 6371.0
        val dLat = Math.toRadians(KAABA_LAT - latitude)
        val dLng = Math.toRadians(KAABA_LNG - longitude)
        val a = sin(dLat / 2) * sin(dLat / 2) +
            cos(Math.toRadians(latitude)) * cos(Math.toRadians(KAABA_LAT)) *
            sin(dLng / 2) * sin(dLng / 2)
        val c = 2 * atan2(kotlin.math.sqrt(a), kotlin.math.sqrt(1 - a))
        return earthRadius * c
    }
}

File: iosApp/Saat/Domain/Models/QiblaCalculator.swift
//
//  QiblaCalculator.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

struct QiblaCalculator {
    private static let KAABA_LAT = 21.4225
    private static let KAABA_LNG = 39.8262

    /// Bearing from user location to Kaaba in degrees (0–360, clockwise from north).
    static func bearingToKaaba(latitude: Double, longitude: Double) -> Double {
        let lat1 = latitude * .pi / 180.0
        let lat2 = KAABA_LAT * .pi / 180.0
        let deltaLng = (KAABA_LNG - longitude) * .pi / 180.0

        let y = sin(deltaLng) * cos(lat2)
        let x = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(deltaLng)
        let bearing = atan2(y, x) * 180.0 / .pi
        return (bearing + 360.0).truncatingRemainder(dividingBy: 360.0)
    }

    /// Distance from user location to Kaaba in kilometers.
    static func distanceToKaabaKm(latitude: Double, longitude: Double) -> Double {
        let earthRadius = 6371.0
        let dLat = (KAABA_LAT - latitude) * .pi / 180.0
        let dLng = (KAABA_LNG - longitude) * .pi / 180.0
        
        let a = sin(dLat / 2) * sin(dLat / 2) +
                cos(latitude * .pi / 180.0) * cos(KAABA_LAT * .pi / 180.0) *
                sin(dLng / 2) * sin(dLng / 2)
        let c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return earthRadius * c
    }
}
